src/ir_passes/discharge.rs

```diff
@@ -170,18 +170,21 @@ impl Discharge {
 
     /// Defines primitive functions used in the encoding like `pow` and `log`
     fn define_funcs(&mut self) {
-        let int_sort = self.sol.int_sort();
-        let pow2 = self
-            .sol
-            .declare_fun("pow2", vec![int_sort], int_sort)
-            .unwrap();
-        let log = self
-            .sol
-            .declare_fun("log2", vec![int_sort], int_sort)
-            .unwrap();
-        self.func_map = vec![(ast::UnFn::Pow2, pow2), (ast::UnFn::Log2, log)]
-            .into_iter()
-            .collect();
+        let is = self.sol.int_sort();
+
+        macro_rules! sol_fn(
+            ($name:tt($($args:ident),*) -> $out:ident) => {
+                self.func_map.insert(ast::Fn::$name, self
+                    .sol
+                    .declare_fun(stringify!($name).to_lowercase(), vec![$($args),*], $out)
+                    .unwrap());
+            }
+        );
+
+        sol_fn!(Pow2(is) -> is);
+        sol_fn!(CLog2(is) -> is);
+        sol_fn!(Pow(is, is) -> is);
+        sol_fn!(CLog(is, is) -> is);
```

Comment:
Another issue here: Because the solver can only work on integer values, if we used something like `ceil` in any parameter expression that eventually is used in typechecking, it can lead to broken solver behavior (I think). @rachitnigam is there a way to get around this, or do we need to instead add some checks in the `Validate` pass on whether float-based computations are used anywhere in expressions used by the typechecker?